firedrake/assemble.py

```diff
@@ -80,55 +92,62 @@ def assemble(expr, tensor=None, bcs=None, form_compiler_parameters=None,
-                          diagonal=diagonal,
-                          options_prefix=options_prefix)
-        for l in loops:
-            m = l()
-        return m
+        return assemble_form(expr, *args, **kwargs)
     elif isinstance(expr, ufl.core.expr.Expr):
         return assemble_expressions.assemble_expression(expr)
     else:
         raise TypeError(f"Unable to assemble: {expr}")
 
 
-def get_mat_type(mat_type, sub_mat_type, arguments):
-    """Provide defaults and validate matrix-type selection.
+def assemble_form(expr, tensor=None, bcs=None, *,
+                  diagonal=False,
+                  assembly_type="solution",
+                  form_compiler_parameters=None,
+                  mat_type=None,
+                  sub_mat_type=None,
+                  appctx={},
+                  options_prefix=None,
+                  nest=None):
+    """Assemble an expression.
 
-    :arg mat_type: PETSc matrix type for the assembled matrix.
-    :arg sub_mat_type: PETSc matrix type for blocks if mat_type is
-        nest.
-    :arg arguments: The test and trial functions.
-    :raises ValueError: On bad arguments.
-    :returns: 2-tuple of validated mat_type, sub_mat_type.
+    :arg expr: a :class:`~ufl.classes.Form` or a :class:`~slate.TensorBase`
+        expression.
+
+    See :func:`assemble` for a description of the possible additional arguments.
     """
-    if mat_type is None:
-        mat_type = parameters.parameters["default_matrix_type"]
-        if any(V.ufl_element().family() == "Real"
-               for arg in arguments
-               for V in arg.function_space()):
-            mat_type = "nest"
-    if mat_type not in {"matfree", "aij", "baij", "nest", "dense"}:
-        raise ValueError(f"Unrecognised matrix type, '{mat_type}'")
-    if sub_mat_type is None:
-        sub_mat_type = parameters.parameters["default_sub_matrix_type"]
-    if sub_mat_type not in {"aij", "baij"}:
-        raise ValueError(f"Invalid submatrix type, '{sub_mat_type}' (not 'aij' or 'baij')")
-    return mat_type, sub_mat_type
+    if nest:
+        raise ValueError("Can't use 'nest', set 'mat_type' instead")
+
+    # Do some setup of the arguments and wrap them in a namedtuple.
+    bcs = solving._extract_bcs(bcs)
+    if assembly_type == "solution":
+        assembly_type = _AssemblyType.SOLUTION
+    elif assembly_type == "residual":
+        assembly_type = _AssemblyType.RESIDUAL
+    else:
+        raise ValueError("assembly_type must be either 'solution' or 'residual'")
+    mat_type, sub_mat_type = _get_mat_type(mat_type, sub_mat_type,
+                                           expr.arguments())
+    opts = _AssemblyOpts(diagonal, assembly_type, form_compiler_parameters,
+                         mat_type, sub_mat_type, appctx, options_prefix)
+
+    assembly_rank = _get_assembly_rank(expr, diagonal)
+    if assembly_rank == _AssemblyRank.SCALAR:
+        return _assemble_scalar(expr, tensor, bcs, opts)
```

Comment:
Given that your documentation for `_assemble_scalar` implies that `tensor` must be `None` - won't this fail if `tensor` is not `None` in kwargs of `assemble`?

Reply:
I've moved the check for `tensor` being `None` to here so it should be clearer